### PR TITLE
chore(medusa): Simplify/improve orders list and count

### PR DESCRIPTION
**Pull request by adrien2p**
FIXES CORE-1414

### CODE REVIEW

**Comment by adrien2p** [comment]
@olivermrbl what do you think for this one? Would it be reasonable to create a new pr with all the changes except the migrations?

**Comment by olivermrbl** [comment]
> @olivermrbl what do you think for this one? Would it be reasonable to create a new pr with all the changes except the migrations?


Yes, think it makes sense to separate the two :)

File: packages/medusa/src/models/customer.ts [hunk 1]
```diff
@@ -20,6 +20,8 @@ import { generateEntityId } from "../utils/generate-entity-id"
 
 @Entity()
 @Unique(["email", "has_account"])
+@Index("idx_customer_first_last_phone", ["first_name", "last_name", "phone"])
```
Comment (carlos-r-l-rodrigues) [inline packages/medusa/src/models/customer.ts]: Why do we have an index for these fields?

Comment (adrien2p) [reply packages/medusa/src/models/customer.ts]: Nice catch, i need to remove this one I thought I did it 👍

File: packages/medusa/src/migrations/1699564794650-update-order-customer-address-indexes.ts [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+      CREATE INDEX IF NOT EXISTS idx_gin_customer_phone ON customer USING gin (phone gin_trgm_ops);
+      CREATE INDEX IF NOT EXISTS idx_gin_customer_first_name ON customer USING gin (first_name gin_trgm_ops);
+      CREATE INDEX IF NOT EXISTS idx_gin_customer_last_name ON customer USING gin (last_name gin_trgm_ops);
+      `
+      )
+    } catch (e) {
+      // noop
+      // The extension might not be installed, in that case do nothing except warn
+      console.warn(
+        "Could not create pg_trgm extension or indexes, skipping. If you want to use the pg_trgm extension, please install it manually and then run the migration productSearchGinIndexes1679950645254.",
```
Comment (olivermrbl) [inline packages/medusa/src/migrations/1699564794650-update-order-customer-address-indexes.ts]:
```suggestion
        "Could not create pg_trgm extension or indexes, skipping. If you want to use the pg_trgm extension, please install it manually and then run the migration updateOrderCustomerAddressIndexes1699564794650.",
```

File: packages/medusa/src/models/customer.ts (hunk 1 above)
Comment (olivermrbl) [reply packages/medusa/src/models/customer.ts]: todo:

File: packages/medusa/src/services/order.ts [hunk 3]
```diff
@@ -301,10 +284,9 @@ class OrderService extends TransactionBaseService {
 
     delete query.relations
 
-    const raw = await orderRepo.findWithRelations(rels, query)
-    const count = await orderRepo.count(query)
-    const orders = await promiseAll(
-      raw.map(async (r) => await this.decorateTotals(r, totalsToSelect))
+    let [orders, count] = await orderRepo.findWithRelationsAndCount(rels, query)
+    orders = await Promise.all(
```
Comment (olivermrbl) [inline packages/medusa/src/services/order.ts]: todo: `promiseAll` util

**Review by olivermrbl — CHANGES_REQUESTED** [review]
LGTM, please address previous comment and make the tests pass :)

**Review by olivermrbl — APPROVED** [review]
LGTM, would be great with benchmarks to include in the release notes.

**Comment by adrien2p** [comment]
Yes I will add a benchmark on this pr too

**Comment by adrien2p** [comment]
To perform the tests the following parameters are generated randomly
- q
- limit
- offset
Overall we can see that the number of requests to perform the tests take less time to be done and overall the perf are 3 times better on median

Screenshot 2023-12-19 at 14 09 25
Screenshot 2023-12-19 at 14 16 20

**Comment by adrien2p** [comment]
@olivermrbl I finally found time to move forward on this one. I suspect the tests will fail and I will handle them later. From the benchmark we can see some nice improvements

**Comment by olivermrbl** [comment]
This is great @adrien2p, thx 🏎️

**Pull request closed without merging**